dlup/writers.py

```diff
@@ -0,0 +1,209 @@
+        tile_size : Tuple[int, int]
+            Tiff tile_size, defined as (height, width).
+        pyramid : bool
+            Whether to write a pyramidal image.
+        compression : TiffCompression
+            Compressor to use.
+        quality : int
+            Quality in case a lossy compressor is used.
+        """
+        self._filename = filename
+        self._tile_size = tile_size
+        self._size = (*size, 1) if len(size) == 2 else size
+
+        if isinstance(mpp, float):
+            mpp = (mpp, mpp)
+        self._mpp: Tuple[float, float] = mpp
+
+        if not compression:
+            compression = TiffCompression.NONE
+
+        self._compression = compression
+        self._pyramid = pyramid
+        self._quality = quality
+
+    def from_pil(self, pil_image: PIL.Image):
+        if not np.all(np.asarray(pil_image.size)[::-1] >= self._tile_size):
+            raise RuntimeError(
+                f"PIL Image must be larger than set tile size. Got {pil_image.size} and {self._tile_size}."
+            )
+        iterator = _pil_grid_iterator(pil_image, self._tile_size)
+        self.from_iterator(iterator, filename=self._filename)
+
+    def from_iterator(self, iterator: Iterator[np.ndarray | None]):
+        filename = pathlib.Path(self._filename)
+        temp_filename = filename.with_suffix(f"{filename.suffix}.partial")
+
+        native_size = self._size[:-1]
+        software = f"dlup {dlup.__version__} with tifffile.py backend"
+        n_subresolutions = 0
+        if self._pyramid:
+            n_subresolutions = int(np.ceil(np.log2(np.asarray(native_size) / np.asarray(self._tile_size))).min())
+        shapes = [
+            np.floor(np.asarray(native_size) / 2**n).astype(int).tolist() for n in range(0, n_subresolutions + 1)
+        ]
+
+        native_resolution = 1 / np.array(self._mpp) * 10000
+        metadata = {
+            # "axes": "TCYXS",
+            # "SignificantBits": 10,
+            "PhysicalSizeX": self._mpp[0],
+            "PhysicalSizeXUnit": "µm",
+            "PhysicalSizeY": self._mpp[1],
+            "PhysicalSizeYUnit": "µm",
+        }
+
+        _compression = TIFFFILE_COMPRESSION[self._compression.value]
+
+        is_rgb = self._size[-1] in (3, 4)
+        with tifffile.TiffWriter(temp_filename, bigtiff=True) as tiff_writer:
+            tiff_writer.write(
+                iterator,  # noqa
+                shape=(*shapes[0], self._size[-1]) if is_rgb else (*shapes[0], 1),
+                dtype="uint8",
+                subifds=None,
+                resolution=(*native_resolution, "CENTIMETER"),  # noqa
+                metadata=metadata,
+                photometric="rgb" if is_rgb else "minisblack",
+                compression=_compression if not self._quality else (_compression, self._quality),
+                tile=self._tile_size,
+                software=software,
+            )
+
+            for level in range(0, n_subresolutions):
```

Comment:
just a stylistic choice, but this block is a bit too nested.

Reply:
Fixed